### PR TITLE
Optimize Kotlin Native compiler to reduce total CPU time

### PR DESCRIPTION
**Pull request by MarkCMann**
Profiling internal applications revealed a couple of hotspots within the kotlin native compiler. 

^KT-61121 Fixed

### CODE REVIEW

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt [hunk 1]
```diff
@@ -28,10 +29,18 @@ internal abstract class LlvmModuleSpecificationBase(protected val cachedLibrarie
...
             packageFragment.konanLibrary.let { it == null || containsLibrary(it) }
 
-    override fun containsDeclaration(declaration: IrDeclaration): Boolean =
-            declaration.konanLibrary.let { it == null || containsLibrary(it) }
-                    // When producing per-file caches, some declarations might be generated by the stub generator.
-                    && declaration.getPackageFragment() !is IrExternalPackageFragment
+    private val containsCache = mutableMapOf<IrDeclaration, Boolean>()
+
+    override fun containsDeclaration(declaration: IrDeclaration): Boolean = containsCache.getOrPut(declaration) {
+        val metadata = ((declaration as? IrMetadataSourceOwner)?.metadata as? KonanMetadata)
```
Comment (madsager) [inline kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:35]: This drops the check that the declarations package fragment is not an `IrExternalPackageFragment` and the comment explaining why that is needed. Is that on purpose?

**Review by madsager — COMMENTED** [review]
I have posted one question on an external package fragment check that I think got removed in these changes. Other than that, changes look good to me.

@SvyatoslavScherbina, who would be the right people to review this?

**Comment by SvyatoslavScherbina** [comment]
> @SvyatoslavScherbina, who would be the right people to review this?

I've added two reviewers. Thanks!

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt [hunk 2]
```diff
@@ -28,10 +29,18 @@ internal abstract class LlvmModuleSpecificationBase(protected val cachedLibrarie
     override fun containsPackageFragment(packageFragment: IrPackageFragment): Boolean =
             packageFragment.konanLibrary.let { it == null || containsLibrary(it) }
 
-    override fun containsDeclaration(declaration: IrDeclaration): Boolean =
-            declaration.konanLibrary.let { it == null || containsLibrary(it) }
-                    // When producing per-file caches, some declarations might be generated by the stub generator.
-                    && declaration.getPackageFragment() !is IrExternalPackageFragment
+    private val containsCache = mutableMapOf<IrDeclaration, Boolean>()
+
+    override fun containsDeclaration(declaration: IrDeclaration): Boolean = containsCache.getOrPut(declaration) {
```
Comment (homuroll) [inline kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:34]: What is the purpose of inlining of `declaration.konanLibrary` here?

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt (hunk 1 above)
Comment (MarkCMann) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:35]: Ah missed that, I'll add that back in.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt (hunk 2 above)
Comment (MarkCMann) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:34]: To memoize more of the calls, by inlining the konan library recursive calls here we can store the result for the top level class in addition to all nested classes. 

Overall this method showed up on the profiler as being very heavy due to the repeated recursive calls in KonanLibrary -- I opted to memoize the data in the llvm module specification class instead of trying to add it to the IrDeclaration or adding a global cache to konanLibrary in IrUtils.

Comment (homuroll) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:34]: Got it. Please, add a comment here explaining this, and probably some short comment/reference to here from `declaration.konanLibrary` (in case someone decides to change it).

**Comment by semoro** [comment]
@udalov Please take a look at the changes in
```
  core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt
  core/metadata/src/org/jetbrains/kotlin/metadata/deserialization/NameResolverImpl.kt
```

File: core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt [hunk 3]
```diff
@@ -6,11 +6,18 @@
...
 import org.jetbrains.kotlin.metadata.deserialization.NameResolver
+import org.jetbrains.kotlin.metadata.deserialization.NameResolverImpl
 import org.jetbrains.kotlin.name.ClassId
 import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.name.FqName
 
 fun NameResolver.getClassId(index: Int): ClassId {
-    return ClassId.fromString(getQualifiedClassName(index), isLocalClassName(index))
+    if (this is NameResolverImpl) {
+        val identifiers = traverseIds(index)
```
Comment (udalov) [inline core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt:16]: I'd prefer to add a method `fun getClassId(index: Int): String` to `NameResolver` instead, so that most of the other usages (which create a `ClassId` anyway) would also work more efficiently.

Since fixing most call sites of `getQualifiedClassName` would involve many changes, maybe it'd be better to extract such change into a separate PR.

**Review by udalov — CHANGES_REQUESTED** [review]

File: core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt (hunk 3 above)
Comment (ddolovov) [reply core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt:16]: `ClassId` is in `kotlin.core.compiler.common` module. And `NameResolver` is in `kotlin.core.metadata`. The latter one does not depend on the first one. @udalov Do you think it might be more relevant to do it like described [here](https://github.com/JetBrains/kotlin/pull/5156#discussion_r1265343244) or to fix somehow module dependencies?

File: compiler/ir/serialization.common/src/org/jetbrains/kotlin/backend/common/overrides/FakeOverrides.kt [hunk 4]
```diff
@@ -286,10 +286,9 @@ class FakeOverrideBuilder(
     }
 
     fun provideFakeOverrides() {
-        val entries = fakeOverrideCandidates.entries
+        val entries = fakeOverrideCandidates.entries.toMutableList()
```
[the remark this answers is not in the document]
Comment (MarkCMann) [reply compiler/ir/serialization.common/src/org/jetbrains/kotlin/backend/common/overrides/FakeOverrides.kt:289]: Ah okay it wasn't obvious to me if the entry set returned was the actual underlying entryset for the map. I can add a call to clear the fakeOverrideCandidates here

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/StaticInitializersOptimization.kt [hunk 5]
```diff
@@ -179,12 +179,12 @@ internal object StaticInitializersOptimization {
             val functionsRequiringGlobalInitializerCall = collectFunctionsRequiringInitializerCall(
                     initializedFiles.beforeCallGlobal,
                     callSitesRequiringGlobalInitializerCall.map { it.actualCallee }
-                            .toMutableSet().intersect(callSitesNotRequiringGlobalInitializerCall.map { it.actualCallee })
+                            .toMutableSet().intersect(callSitesNotRequiringGlobalInitializerCall.map { it.actualCallee }.toSet())
```
[the remark this answers is not in the document]
Comment (MarkCMann) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/StaticInitializersOptimization.kt:182]: Agree on the to mutable set call, should I remove it?

I tried to change as little code as possible to highlight the actual hot spots

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt [hunk 6]
```diff
@@ -1785,8 +1785,8 @@ private fun ObjCExportCodeGenerator.createDirectAdapters(
         )
     }
 
-    val inheritedAdapters = superClass?.getAllRequiredDirectAdapters().orEmpty()
-    val requiredAdapters = typeDeclaration.getAllRequiredDirectAdapters() - inheritedAdapters
+    val inheritedAdapters = superClass?.getAllRequiredDirectAdapters().orEmpty().toSet()
+    val requiredAdapters = typeDeclaration.getAllRequiredDirectAdapters().filter { !inheritedAdapters.contains(it) }
```
[the remark this answers is not in the document]
Comment (MarkCMann) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt:1789]: So I need to convert to set to make the contains call fast, otherwise it's essentially calling contains on a list in a loop. 

If your question is more about why did I use filter instead of just calling minus, I'm not sure. I could just move back to calling minus on the set instead of a hand written filter.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt (hunk 2 above)
Comment (MarkCMann) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:36]: @ddolovov I'm not sure that I understand your concern here. Is this more related to the size of the cache? 

With this change I've moved the recursive calls into the llvm contains declaration function so now that function should be called multiple times for `IrDeclarations`.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (ddolovov) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt]: > If your question is more about why did I use filter instead of just calling minus, I'm not sure. I could just move back to calling minus on the set instead of a hand written filter.

Yes, it was exactly about this. I was not sure if you did this on purpose and what purpose.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt (hunk 2 above)
Comment (ddolovov) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:36]: > so now that function should be called multiple times for IrDeclarations.

You mean it could be called multiple times for the same instance of `IrDeclaration` (and therefore it makes sense to cache per-declaration), right?

Comment (MarkCMann) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt:36]: That's correct, especially for parent IrDeclarations they would be called repeatedly

File: core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt [hunk 7]
```diff
@@ -6,11 +6,18 @@
...
 
 import org.jetbrains.kotlin.metadata.deserialization.NameResolver
+import org.jetbrains.kotlin.metadata.deserialization.NameResolverImpl
 import org.jetbrains.kotlin.name.ClassId
 import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.name.FqName
 
 fun NameResolver.getClassId(index: Int): ClassId {
-    return ClassId.fromString(getQualifiedClassName(index), isLocalClassName(index))
+    if (this is NameResolverImpl) {
```
[the remark this answers is not in the document]
Comment (MarkCMann) [reply core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt:15]: Digging into this a bit more, there's a [JVM specific name resolver](https://github.com/JetBrains/kotlin/blob/8820867341b8844f73e1eade1d2a3ce28fa67209/core/metadata.jvm/src/org/jetbrains/kotlin/metadata/jvm/deserialization/JvmNameResolverBase.kt#L61) that optimizes the case of multiple queries better by caching the `index -> isLocal` mapping. Do you want me to deprecate the isLocal function for all NameResolvers?

[the remark this answers is not in the document]
Comment (udalov) [reply core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt:15]: Yes, I would even prefer to remove isLocal function from the interface. I.e. replace these two functions in `NameResolver`:
```
    /**
     * @return the fully qualified name of some class in the format: `org/foo/bar/Test.Inner`
     */
    fun getQualifiedClassName(index: Int): String

    fun isLocalClassName(index: Int): Boolean
```
with:
```
    fun getClassName(index: Int): ClassName

    /**
     * @param name the fully qualified name of some class in the format: `org/foo/bar/Test.Inner`
     */
    class ClassName(val name: String, val isLocal: Boolean)
```
and **maybe** an extension (not sure about it):
```
fun NameResolver.getQualifiedClassName(index: Int): String = getClassName(index).name
```

But I would suggest to do that in a separate PR.

[the remark this answers is not in the document]
Comment (MarkCMann) [reply core/deserialization.common/src/org/jetbrains/kotlin/serialization/deserialization/NameResolverUtil.kt]: Alright I've removed the NameResolver changes from this PR and will put up a second PR later with the NameResolver changes.

**Comment by MarkCMann** [comment]
Hey @udalov, I think that I've addressed all your outstanding concerns. Can you take another look?

**Comment by udalov** [comment]
^ I think it's rather for @ddolovov. (I was only reviewing the NameResolver changes, which were extracted.)

**Comment by ddolovov** [comment]
@MarkCMann Please create a YouTrack issue with the description for that change as required by the [Contribution Guide](https://github.com/JetBrains/kotlin/blob/master/docs/contributing.md#rules-for-commit-messages)

**Comment by ddolovov** [comment]
> @MarkCMann Please create a YouTrack issue with the description for that change as required by the [Contribution Guide](https://github.com/JetBrains/kotlin/blob/master/docs/contributing.md#rules-for-commit-messages)

JFTR: https://youtrack.jetbrains.com/issue/KT-61121

**Comment by ddolovov** [comment]
Merged.
